sdcm/mgmt.py

```diff
@@ -1,13 +1,408 @@
+# coding: utf-8
+from avocado.utils import process
 import requests
 import logging
 import json
 import time
 import datetime
 
+from sdcm import wait
+
 logger = logging.getLogger(__name__)
 
 STATUS_DONE = 'done'
 STATUS_ERROR = 'error'
+MANAGER_IDENTITY_FILE = '/tmp/scylla_manager_pem'
+
+from enum import Enum
+
+class ScyllaManagerError(Exception):
+    """
+    A custom exception for Manager related errors
+    """
+    pass
+
+
+class TaskStatus(Enum):
+   NEW = "NEW"
+   RUNNING = "RUNNING"
+   DONE = "DONE"
+   UNKNOWN = "UNKNOWN"
+   ERROR = "ERROR"
+   STOPPED = "STOPPED"
+
+   @classmethod
+   def from_str(cls, output_str):
+       try:
+            output_str = output_str.upper()
+            return getattr(cls, output_str)
+       except AttributeError:
+           raise ScyllaManagerError("Could not recognize returned task status: {}".format(output_str))
+
+
+class ScyllaLogicObj(object):
```

Comment:
perhaps this sketch will help clarify this design: [scylla-manager model sketch](https://drive.google.com/file/d/15oC1yFhwyLTSQTaGaVPvorwXaDlGBl1M/view?ths=true)

- ScyllaLogicObj provides parsing of the sctool output **inherently relies on the object id**.
the "id" property is what allows all functions not to send and receive hundreds of cluster/task id on every call. for example - to write a test command of: " mgr_repair_task.stop() ".
(see this line in function: `filter_param_updated = filter_param or self.id`)
- the "run_sctool_cmd" is on the contrary nothing but an interface, which can be seem like an alternative to "run_rest_api".
so these two "runners" are abilities of the manager-tool (see in sketch). while the other objects of "cluster" and "task" are a representation of the Manager object model (using the common unique identifier of "id").
you're right that the design is not fully object-oriented. that is because i didn't want to rewrite all the REST API related code as well. so perhaps about ~ 75% of code is now OO. it may be a good idea to rewrite REST part as well, so that, looking at the sketch - the "scylla manager tool" can become a base class for all others.

Reply:
I'll try to explain what I mean:

`
class ScyllaManagerCLIBase(object):

    def run_sctool_cmd(self, cmd):
        pass

    def _get_output_table_value(self):
        pass

class ScyllaClusterManager(ScyllaManagerCLIBase):

    def __init__(self):
        self.id = None

    def create(self):
        self.run_sctool_cmd("...")
        cluster_id = ...  # get cluster_id
        self.id = cluster_id

    def get(self):
        if not self.id:
            self.run_sctool_cmd("...")
            cluster_id = 0  # get cluster_id
            self.id = cluster_id
        return self.id

    def create_repair_task(self):
        self.run_sctool_cmd("...")
        task_id = ... # get task_id
        return ScyllaTaskManager(task_id, self.id)


class ScyllaTaskManager(ScyllaManagerCLIBase):

    def __init__(self, task_id, cluster_id):
        self.id = task_id
        self.cluster_id = cluster_id

class ScyllaManager(object):

    def create_cluster(self):
        cm = ScyllaClusterManager()
        cm.create()
        return cm
`

Reply:
implemented changes (along with added health-check support) of:
* 28cc2da - (HEAD -> multidc-yaml, origin/multidc-yaml, healthcheck) 
(1) Added new class SCTool to contain all sctool related functions. 
(2) Added SCTool object to base class of ScyllaLogicObj. 
(3) removed passing base object self parameter to other objects.